ui-speedspacechart/src/components/helpers/drawElements/speedLimitTags.ts

```diff
@@ -82,108 +85,135 @@ export const drawSpeedLimitTags = async ({
-          }
-        } else {
-          ctx.fillStyle = 'white';
-          ctx.font = '600 12px "IBM Plex Sans"';
-          ctx.textAlign = 'left';
-          ctx.textBaseline = 'middle';
-          const textPosition = x + TEXT_LEFT_PADDING;
-          ctx.fillText(tag, textPosition, Y_POSITION + TEXT_PADDING_TOP + RECT_HEIGHT / 2);
-        }
+        const iconXPosition = x + (tagWidth - ICON_BACKGROUND_WIDTH) / 2;
+        const iconYPosition = Y_POSITION - ICON_OFFSET;
+        const cornerRadius = 4;
+
+        drawRoundedRect(
+          ctx,
+          iconXPosition,
+          iconYPosition,
+          ICON_BACKGROUND_WIDTH,
+          ICON_BACKGROUND_HEIGHT,
+          cornerRadius
+        );
+
+        drawSvgImageWithColor(
+          ctx,
+          image,
+          iconXPosition + ICON_OFFSET,
+          iconYPosition + ICON_OFFSET,
+          ICON_WIDTH,
+          ICON_HEIGHT,
+          '#FFFFFF'
+        );
+      } else {
+        ctx.fillStyle = 'white';
+        ctx.font = '600 12px "IBM Plex Sans"';
+        ctx.textAlign = 'left';
+        ctx.textBaseline = 'middle';
+        const textPosition = x + TEXT_LEFT_PADDING;
+        ctx.fillText(tag, textPosition, Y_POSITION + TEXT_PADDING_TOP + RECT_HEIGHT / 2);
       }
-    });
+    }
+  });
 
-    drawSeparatorLinearLayer(ctx, 'rgba(0,0,0,0.1)', MARGINS, width, marginTop);
-    ctx.restore();
+  drawSeparatorLinearLayer(ctx, 'rgba(0,0,0,0.1)', MARGINS, width, marginTop);
+  ctx.restore();
 
-    // prevent overlapping with margins left and right
-    ctx.clearRect(0, 0, MARGIN_LEFT, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
-    ctx.clearRect(width - MARGIN_RIGHT, 0, width, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+  // prevent overlapping with margins left and right
+  ctx.clearRect(0, 0, MARGIN_LEFT, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+  ctx.clearRect(width - MARGIN_RIGHT, 0, width, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+};
+
+export const computeTooltip = async ({
+  width,
+  height: marginTop,
+  store,
+}: DrawFunctionParams): Promise<tooltipInfos | null> => {
+  const { speedLimitTags, ratioX, leftOffset, cursor } = store;
+
+  const { MARGIN_TOP, MARGIN_LEFT } = MARGINS;
+
+  const maxPosition = maxPositionValue(store.speeds);
+
+  let tooltip: tooltipInfos | null = null;
 
-    return tooltip;
+  if (speedLimitTags == undefined) return null;
+
+  for (const { position, value } of speedLimitTags) {
+    const tooltipTextMap: Record<string, string> = {
+      incompatible: 'Incompatible with the infrastructure',
+    };
+
+    const { tag } = value;
+    if (!(tag in tooltipTextMap)) continue;
+    const potential_tooltip_text = tooltipTextMap[tag];
+
+    const x = positionOnGraphScale(position.start, maxPosition, width, ratioX, MARGINS);
+    const nextBoundary = positionOnGraphScale(position.end!, maxPosition, width, ratioX, MARGINS);
+
+    if (nextBoundary == null || cursor.x == null || cursor.y == null) break;
```

Comment:
Since == null matches exactly null and undefined and nothing else, this is actually the only case where I prefer == to ===. So I would rather keep it like this unless you insist.

Reply:
you're right, I just prefer to opt for the most restrictive when it's possible, it seems safer to me, but I don't want to look pyschorigid so as you wish

Reply:
We try to avoid `==` because it has quite weird rules. In general, I go for `if (!thing)` instead of `if (thing == null)` (but `if (thing === null)` is fine by me as well).

Reply:
another way
```suggestion
    if (!nextBoundary || !cursor.x || !cursor.y) break;
```

Reply:
No, !x would be different here, because these are numbers, so x == null test if the value is nullish (only null and undefined) but !x tests if it's falsy, which includes values like 0

The equivalent would be x === null || x === undefined

Reply:
ah yes of course, my bad

Reply:
Tbf I did this to avoid caring about whether they could be null or undefined, especially since in the previous version I think the type of one of them appeared like it could be neither null nor undefined, despite being tested. So perhaps I should actually investigate it instead of opting for the lazy way '^^

Reply:
Yet another way, if you prefer:

```ts
typeof cursor.x === 'number'
```